File: Python/Code/Tree/1-6.py
# -*- coding: UTF-8 -*-
from tree import Tree
import sys
import os
import math
# 根据有序数组重建平衡搜索二叉树,默认递增
def generateTree(arr):
    return headFromArr(arr, 0, len(arr) - 1)
    
def headFromArr(arr, begin, end):
    if begin > end:
        return None
    index = int(math.ceil((end+begin)/2))
    node = Tree(arr[index])
    node.left = headFromArr(arr, begin, index - 1)
    node.right = headFromArr(arr, index + 1, end)
    return node
def middleTraverse(node):
    if node == None:
        return
    middleTraverse(node.left)
    print(node.value)
    middleTraverse(node.right)

if __name__ == "__main__":
    node = generateTree([1,2,3,4,5,6,7])
    middleTraverse(node)

File: JavaScript/小浩算法/动态规划-DP/16.js
/*
三角形的最小路径和
*/

function soluation(arr) {
    let res = []

    for (let i = 0; i < arr.length; i++) {
        const elements = arr[i];
        let temArr = []

        for (let j = 0; j < elements.length; j++) {
            const element = elements[j];
            if (res.length == 0) {
                temArr[j] = element
            } else {
                let left = j - 1
                let right = j
                if (left < 0) { left = 0 }
                if (right > res.length - 1) { right = res.length - 1 }
                temArr[j] = Math.min(res[left], res[right]) + element
            }
        }
        res = temArr
    }
    return res
}
let arr = [
    [2],
    [3, 4],
    [10, 10, 10],
    [4, 1, 8, 5],
    [5, 20, 6, 2, 0]
]
console.log(soluation(arr))